src/dict.c

```diff
@@ -588,6 +602,8 @@ static dictEntry *dictGenericDelete(dict *d, const void *key, int nofree) {
                     dictFreeUnlinkedEntry(d, he);
                 }
                 d->ht_used[table]--;
+                if (_dictShrinkIfNeeded(d) == DICT_ERR) 
```

Comment:
I believe we should not and do not need to check the return value of `_dictShrinkIfNeeded`. Regardless of whether it triggers a resize, we should return `he` to the caller.

The same applies to `dictFindPositionForInsert`. We should not check if `_dictExpandIfNeeded` actually triggers a resize. Whether a resize happens or not is not directly related to insertions or deletions. However, this check has been present in Redis since its inception. Does anyone know the history behind this? I'm afraid only antirez knows...

Reply:
i agree, let's change it for both expand and shrink.

looking at redis 2.8 code, i think that case is not reachable.
the flow is dictAddRaw->_dictExpandIfNeeded->dictExpand
and the error in dictExpand is triggered in a condition that's first covered by _dictExpandIfNeeded (which returns success in that case)

Reply:
I just ignore the return value but I'm not sure if I should change the return value of `_dictShrinkIfNeeded` and `_dictExpandIfNeeded` to void because it may be used in the future.

Reply:
let's keep the return value, and just avoid checking it.

also, let's mention this change and the reason for it in the top comment.
that is if @soloestoy agrees with my previous comment.

Reply:
I hope we can go further :-). Taking `_dictExpandIfNeeded` and `dictExpand` as examples:

```c
/* return DICT_ERR if expand was not performed */
int dictExpand(dict *d, unsigned long size) {
    return _dictExpand(d, size, NULL);
}
```
The return value of `dictExpand` is described very clearly: "return DICT_ERR if expand was not performed", and its execution logic is also consistent with the description.

However, the logic of `_dictExpandIfNeeded` is contradictory:
```c
/* Expand the hash table if needed */
static int _dictExpandIfNeeded(dict *d)
{
    /* Incremental rehashing already in progress. Return. */
    if (dictIsRehashing(d)) return DICT_OK;

    /* If the hash table is empty expand it to the initial size. */
    if (DICTHT_SIZE(d->ht_size_exp[0]) == 0) return dictExpand(d, DICT_HT_INITIAL_SIZE);

    /* If we reached the 1:1 ratio, and we are allowed to resize the hash
     * table (global setting) or we should avoid it but the ratio between
     * elements/buckets is over the "safe" threshold, we resize doubling
     * the number of buckets. */
    if ((dict_can_resize == DICT_RESIZE_ENABLE &&
         d->ht_used[0] >= DICTHT_SIZE(d->ht_size_exp[0])) ||
        (dict_can_resize != DICT_RESIZE_FORBID &&
         d->ht_used[0] / DICTHT_SIZE(d->ht_size_exp[0]) > dict_force_resize_ratio))
    {
        if (!dictTypeExpandAllowed(d))
            return DICT_OK;
        return dictExpand(d, d->ht_used[0] + 1);
    }
    return DICT_OK;
}
```
We can see that if `dictExpand` is not actually called (i.e., no resize is triggered), it returns `DICT_OK`. However, the return value of `dictExpand` conflicts with it. I think the return value of `_dictExpandIfNeeded` should be consistent with `dictExpand`, or simply not have a return value at all.

Reply:
ok, so in that case let's just remove the return value from the "IfNeeded" version. whoever calls this function shouldn't care what it decided to do.

Reply:
@soloestoy what about mentioning this change (the fact expand failure doesn't abort the insertion) in the top comment.
did you see what i wrote here: https://github.com/redis/redis/pull/12850#discussion_r1441418714

Reply:
I don't think we need to explain it as it doesn't have any practical impact.

Reply:
you mean because all the conditions in _dictExpand that led to returning DICT_ERR should be unreachable in this flow?
ok...